src/garage/torch/algos/dqn.py

```diff
@@ -0,0 +1,268 @@
+                                            module=self._qf,
+                                            lr=qf_lr)
+        self._eval_env = eval_env
+        self.worker_cls = FragmentWorker
+
+    def train(self, trainer):
+        """Obtain samplers and start actual training for each epoch.
+
+        Args:
+            trainer (Trainer): Experiment trainer.
+
+        Returns:
+            float: The average return in last epoch cycle.
+
+        """
+        if not self._eval_env:
+            self._eval_env = trainer.get_env_copy()
+        last_returns = [float('nan')]
+        trainer.enable_logging = False
+
+        for _ in trainer.step_epochs():
+            for cycle in range(self._steps_per_epoch):
+                trainer.step_path = trainer.obtain_episodes(trainer.step_itr)
+                self.train_once(trainer.step_itr, trainer.step_path)
+                if (cycle == 0 and self.replay_buffer.n_transitions_stored >=
+                        self._min_buffer_size):
+                    trainer.enable_logging = True
+                    logger.log('Evaluating policy')
+
+                    epsilon_before = self.exploration_policy.epsilon
+                    eval_eps = obtain_evaluation_episodes(
+                        self.exploration_policy,
+                        self._eval_env,
+                        num_eps=10,
+                        max_episode_length=self._max_episode_length_eval)
+                    self.exploration_policy.epsilon = epsilon_before
+
+                    last_returns = log_performance(trainer.step_itr,
+                                                   eval_eps,
+                                                   discount=self._discount)
+                    self._episode_reward_mean.extend(last_returns)
+                    tabular.record('Evaluation/100EpRewardMean',
+                                   np.mean(self._episode_reward_mean))
+                trainer.step_itr += 1
+
+        return np.mean(last_returns)
+
+    def train_once(self, itr, episodes):
+        """Perform one iteration of training.
+
+        Args:
+            itr (int): Iteration number.
+            episodes (EpisodeBatch): Batch of episodes.
+
+        """
+        self.replay_buffer.add_episode_batch(episodes)
+
+        epoch = itr / self._steps_per_epoch
+
+        for _ in range(self._n_train_steps):
+            if (self.replay_buffer.n_transitions_stored >=
+                    self._min_buffer_size):
+                samples = self.replay_buffer.sample_transitions(
+                    self._buffer_batch_size)
+                samples['rewards'] *= self._reward_scale
+                qf_loss, y, q = tuple(v.cpu().numpy()
+                                      for v in self.optimize_qf(samples))
+
+                self._episode_qf_losses.append(qf_loss)
+                self._epoch_ys.append(y)
+                self._epoch_qs.append(q)
+
+        if itr % self._steps_per_epoch == 0:
```

Comment:
ideally, can you break this out into a separate logging function?

Reply:
Done.